server/storage/backend/batch_tx.go

```diff
@@ -342,3 +366,182 @@ func (t *batchTxBuffered) UnsafeSeqPut(bucket Bucket, key []byte, value []byte)
 	t.batchTx.UnsafeSeqPut(bucket, key, value)
 	t.buf.putSeq(bucket, key, value)
 }
+
+type batchTxBufferedAsync struct {
+	*batchTxBuffered
+	asyncBuf        txWriteBuffer
+	asyncBufLock    sync.Mutex
+	asyncBufPending int
+}
+
+func newBatchTxBufferedAsync(backend *backend) *batchTxBufferedAsync {
+	twb := txWriteBuffer{
+		txBuffer:   txBuffer{make(map[BucketID]*bucketBuffer)},
+		bucket2seq: make(map[BucketID]bool),
+	}
+	tx := &batchTxBufferedAsync{}
+	tx.batchTxBuffered = newBatchTxBuffered(backend)
+	tx.asyncBuf = twb
+
+	return tx
+}
+
+func (t *batchTxBufferedAsync) safePending() int {
+	t.Mutex.Lock()
+	defer t.Mutex.Unlock()
+	t.asyncBufLock.Lock()
+	defer t.asyncBufLock.Unlock()
+
+	return t.pending + t.asyncBufPending
+}
+
+func (t *batchTxBufferedAsync) LockAsync() {
+	t.asyncBufLock.Lock()
+}
+
+func (t *batchTxBufferedAsync) UnlockAsync() {
+	if t.asyncBufPending != 0 {
+		t.backend.readTx.Lock() // blocks txReadBuffer for writing.
+		t.asyncBuf.writeback(t.backend.readTx.buf)
+		t.backend.readTx.Unlock()
+	}
+	t.asyncBufLock.Unlock()
+}
+
+func (t *batchTxBufferedAsync) Commit() {
+	t.Lock()
+	t.commit(false)
+	t.Unlock()
+}
+
+func (t *batchTxBufferedAsync) CommitAndStop() {
+	t.Lock()
+	t.commit(true)
+	t.Unlock()
+}
+
+func (t *batchTxBufferedAsync) commit(stop bool) {
+	t.commitAndPut(stop)
+}
+
+func (t *batchTxBufferedAsync) commitAndPut(stop bool) {
+	// all read txs must be closed to acquire boltdb commit rwlock
+	t.asyncBufLock.Lock()
+	t.backend.readTx.Lock()
+
+	var index uint64
+	var term uint64
+	if t.backend.hooks != nil {
+		index, term = t.backend.hooks.OnPreCommitIndexAndTermUnsafe()
+	}
+
+	t.pending += t.asyncBufPending
+	t.asyncBufPending = 0
+	t.backend.readTx.committingBuf.reset()
+	bufVersion := t.backend.readTx.buf.bufVersion
+	t.backend.readTx.committingBuf, t.backend.readTx.buf = t.backend.readTx.buf, t.backend.readTx.committingBuf
+	t.backend.readTx.buf.bufVersion = bufVersion
+
+	t.backend.readTx.Unlock()
+	t.asyncBufLock.Unlock()
+
+	if t.backend.hooks != nil {
+		t.backend.hooks.OnPreCommitWithIndexAndTermUnsafe(t, index, term)
+	}
+
+	commitDone := make(chan interface{}, 1)
+
+	go t.unsafeCommitAndPut(stop, commitDone)
```

Comment:
we dont need a channel here, right?

Reply:
We need a channel: https://github.com/boltdb/bolt/issues/392.
Sometimes, bolt.begin(false) will lock bolt.commit() for DB.mmaplock. So we need rollback readTx to help  bolt.commit() get DB.mmaplock.